### PR TITLE
regression tests for fl ops

### PR DESCRIPTION
**Pull request by bennn**
For github issues #111 #113 #114 #115

### CODE REVIEW

File: typed-racket-test/unit-tests/typecheck-tests.rkt [hunk 1]
```diff
@@ -425,6 +425,29 @@
         (tc-e (expt -0.0 -1.0) -NonNegFlonum)
         (tc-e (expt 0 (flabs (cos (real->double-flonum 2))))
               -Real)
+        (tc-err ;; Github issue #115
+          ((lambda ([x : Nonpositive-Flonum]) x)
```
Comment (bennn) [inline typed-racket-test/unit-tests/typecheck-tests.rkt]: At first I tried using `ann`, but then the tests didn't work. Is there a better way to give an incorrect type that I want to assert fails?

For example if I replace `lambda` with `ann` in this test I get:

```
Typechecker tests > tc-expr tests > 429 (FAIL (ann (let ((x : Flonum 5.0)) (if (fl>= x (ann -4.0 Nonpositive-Flonum)) x -3.0)) Nonpositive-Flonum))
429 (FAIL (ann (let ((x : Flonum 5.0)) (if (fl>= x (ann -4.0 Nonpositive-Flonum)) x -3.0)) Nonpositive-Flonum))
FAILURE
location:   <pkgs>/typed-racket-test/unit-tests/typecheck-tests.rkt:428:8
actual:     (tc-results
 (list
  (tc-result
   (U Float-Nan Float-Positive-Zero Float-Negative-Zero Negative-Float-No-NaN)
   (Top | Bot)
   -))
 #f)
expected:   (tc-results (list (tc-result (U) (Bot | Bot) -)) #f)
tc-expr did not return the expected value. The types don't match.
```

**Comment by samth** [comment]
First, you're right that `(ann e T)` always has type `T`; this is an important invariant in Typed Racket. But I don't see why these need to be error tests in the first place. Couldn't we just check that they have the appropriate type? Or if you want to make sure that certain argument combinations don't work, the use them explicitly.

**Comment by bennn** [comment]
Fair point, I had just copied endobson's examples to make the tests. I've gone back and made 1 of the 4 a positive test.

But for the others, I think it's easier to check errors. For issues 111 113 and 115, the correct type is a supertype of the one we want to rule out. So checking that we get a `Flonum` doesn't rule out that the value would typecheck as a `Nonpositive-Flonum` (right?).

**Comment by samth** [comment]
No, the `tc-e` form checks that the result is exactly as specified (try it out).

**Comment by bennn** [comment]
Oh that's nice. But I'm having trouble getting the types to match exactly. (I guess, on Monday I can talk to Asumu and get these working correctly.)

Here I expect `x` to have type `Flonum` exactly (that's what happens in `#lang typed/racket`)
```
(tc-e
 (let ([x : Flonum 5.0])
  (if (fl>= x (ann -4.0 Nonpositive-Flonum))
   x
   -3.0))
 -Flonum)
```

But it's got the wrong filters, I think.

```
actual:     (tc-results
 (list
  (tc-result
   (U
    Float-Nan
    Float-Positive-Zero
    Float-Negative-Zero
    Positive-Float-No-NaN
    Negative-Float-No-NaN)
   (Top | Bot)
   -))
 #f)
expected:   (tc-results
 (list
  (tc-result
   (U
    Float-Nan
    Float-Positive-Zero
    Float-Negative-Zero
    Positive-Float-No-NaN
    Negative-Float-No-NaN)
   (Top | Top)
   -))
 #f)
```

**Comment by stamourv** [comment]
The `actual` filter looks right to me. You can pass a `#:ret` argument to `tc-e` to specify filters in expected results.

**Comment by bennn** [comment]
Thanks @stamourv. They're all positive tests except for [the one about `fl+`](https://github.com/racket/typed-racket/pull/340/files#diff-14aac72a8cf6cd60384e6a05c1c62423R443). Ok?

**Comment by stamourv** [comment]
Sure. That works.
And, as @samth said, the tests check that the actual types are *exactly* the same as the expected ones, so they already catch errors in both directions.
LGTM.

**Comment by samth** [comment]
Looks good to me too.

**Comment by bennn** [comment]
thanks sam

**Pull request merged**